### ISSUE TITLE
[M4.1.2] Train TreeLSTM on contrastive similarity pairs

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#199]
## Parent Issue

#197 - Implement TreeLSTM sentence encoder

## Goal

Train TreeLSTM encoder using contrastive learning on Esperanto sentence similarity pairs.

## Training Strategy

**Contrastive Learning Objective:**
- Positive pairs: Morphological variants (same meaning, different form)
  - "La hundo dormas" ↔ "La hundoj dormis" (same root: hund+dorm)
- Hard negatives: Different roots with similar grammar
  - "La hundo dormas" ⊥ "La kato dormas" (different: hund vs kat)

**Loss Function:**
```python
def contrastive_loss(query, positive, negative, margin=0.5):
    """
    Triplet margin loss: Pull positives closer, push negatives apart.
    """
    pos_sim = cosine_similarity(query, positive)
    neg_sim = cosine_similarity(query, negative)
    
    loss = max(0, margin - pos_sim + neg_sim)
    return loss
```

## Implementation

Create `scripts/train_treelstm.py`:

```python
#!/usr/bin/env python3
"""Train TreeLSTM sentence encoder."""

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
from pathlib import Path
import json

from klareco.models.tree_lstm import TreeLSTMEncoder
from klareco.embeddings.compositional import CompositionalEmbedding
from klareco.ast_to_graph import ASTToGraphConverter
from klareco.parser import parse

class SimilarityTripletDataset(Dataset):
    """Dataset of (query, positive, negative) AST triplets."""
    
    def __init__(
        self,
        corpus_path: Path,
        converter: ASTToGraphConverter,
        num_triplets: int = 100_000,
    ):
        self.corpus_path = corpus_path
        self.converter = converter
        self.triplets = []
        
        # Generate triplets from corpus
        self._generate_triplets(num_triplets)
    
    def _generate_triplets(self, num_triplets):
        """Generate triplets from corpus."""
        print(f"Generating {num_triplets:,} triplets...")
        
        # Load corpus
        sentences = []
        with open(self.corpus_path) as f:
            for line in f:
                doc = json.loads(line)
                sentences.append(doc)
        
        # Group by root words (positive pairs share roots)
        from collections import defaultdict
        root_groups = defaultdict(list)
        
        for sent in sentences:
            ast = sent['ast']
            roots = self._extract_roots(ast)
            root_sig = tuple(sorted(roots))
            root_groups[root_sig].append(sent)
        
        # Generate triplets
        import random
        root_sigs = list(root_groups.keys())
        
        for _ in range(num_triplets):
            # Sample query root signature
            query_sig = random.choice(root_sigs)
            
            # Positive: same roots
            if len(root_groups[query_sig]) < 2:
                continue  # Need at least 2 for positive pair
            
            query_sent, pos_sent = random.sample(root_groups[query_sig], 2)
            
            # Negative: different roots
            neg_sig = random.choice([s for s in root_sigs if s != query_sig])
            neg_sent = random.choice(root_groups[neg_sig])
            
            self.triplets.append({
                'query': query_sent,
                'positive': pos_sent,
                'negative': neg_sent,
            })
        
        print(f"Generated {len(self.triplets):,} triplets")
    
    def _extract_roots(self, ast):
        """Extract root words from AST."""
        roots = set()
        
        def visit(node):
            if isinstance(node, dict):
                if 'radiko' in node:
                    roots.add(node['radiko'])
                for v in node.values():
                    if isinstance(v, (dict, list)):
                        visit(v)
            elif isinstance(node, list):
                for item in node:
                    visit(item)
        
        visit(ast)
        return roots
    
    def __len__(self):
        return len(self.triplets)
    
    def __getitem__(self, idx):
        triplet = self.triplets[idx]
        
        # Convert ASTs to graphs
        query_graph = self.converter.ast_to_graph(triplet['query']['ast'])
        pos_graph = self.converter.ast_to_graph(triplet['positive']['ast'])
        neg_graph = self.converter.ast_to_graph(triplet['negative']['ast'])
        
        return {
            'query': query_graph,
            'positive': pos_graph,
            'negative': neg_graph,
        }

def train_epoch(model, dataloader, optimizer, device):
    """Train for one epoch."""
    model.train()
    total_loss = 0
    
    for batch in dataloader:
        optimizer.zero_grad()
        
        # Encode all three
        query_emb = model(
            batch['query'].x.to(device),
            batch['query'].edge_index.to(device),
        )
        pos_emb = model(
            batch['positive'].x.to(device),
            batch['positive'].edge_index.to(device),
        )
        neg_emb = model(
            batch['negative'].x.to(device),
            batch['negative'].edge_index.to(device),
        )
        
        # Triplet margin loss
        pos_sim = F.cosine_similarity(query_emb, pos_emb)
        neg_sim = F.cosine_similarity(query_emb, neg_emb)
        
        loss = torch.clamp(0.5 - pos_sim + neg_sim, min=0).mean()
        
        loss.backward()
        optimizer.step()
        
        total_loss += loss.item()
    
    return total_loss / len(dataloader)

def main():
    # Load compositional embeddings
    vocab_dir = Path("data/vocabularies")
    comp_embed = CompositionalEmbedding.from_vocabulary_files(vocab_dir)
    
    # Create converter and model
    converter = ASTToGraphConverter(compositional_embedding=comp_embed)
    model = TreeLSTMEncoder(input_dim=128, hidden_dim=128)
    
    # Create dataset
    dataset = SimilarityTripletDataset(
        corpus_path=Path("data/corpus/unified_corpus.jsonl"),
        converter=converter,
        num_triplets=100_000,
    )
    
    dataloader = DataLoader(dataset, batch_size=32, shuffle=True)
    
    # Train
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model.to(device)
    
    for epoch in range(10):
        loss = train_epoch(model, dataloader, optimizer, device)
        print(f"Epoch {epoch+1}: loss = {loss:.4f}")
    
    # Save model
    torch.save(model.state_dict(), "models/treelstm_encoder.pt")

if __name__ == "__main__":
    main()
```

## Acceptance Criteria

- [ ] Training script generates triplets from corpus
- [ ] Positive pairs share roots, negatives don't
- [ ] Contrastive loss pulls positives together
- [ ] Model trains without errors
- [ ] Saved model can be loaded for inference
- [ ] Validation shows improved similarity alignment
